### DISCUSSION TITLE
Cascade delete/restore support for SoftDelete trait

### DISCUSSION THREAD

**Discussion by om3rcitak** [laravel/framework#32933] [General]
Hi,

Laravel do not support cascade delete with soft delete trait at the moment.

There is a 2 package for this job;
- https://github.com/Askedio/laravel-soft-cascade/
- https://github.com/michaeldyrynda/laravel-cascade-soft-deletes

But both 2 packages is not perfect.
- Michael's package is don't support "restore" method, don't have exception classes, don't have listeners etc. So, this package is not usable and stable.
- Askedio's package is;
    - Unecessary complex. For example need to be set "update" action suffix for every relation. Details: https://github.com/Askedio/laravel-soft-cascade/pull/116
    - Not stable and there is a many unfixed issue. https://github.com/Askedio/laravel-soft-cascade/issues

I think this feature is should be in Laravel core. We shouldn't include 3. part library.

**Comment by denisdulici** [comment]
That's why we had to create a small [workaround](https://github.com/akaunting/akaunting/blob/master/app/Traits/Relationships.php#L32). Having this in core would save our lives.

**Comment by netpok** [comment]
In general I try to avoid using cascade on relations as well as using `Model::whereField('something')->delete()` since none of these triggers the events or observers on the model. (On the other hand cascading delete in the db is way more efficient.)

But in this case the main problem with the restoring is how do you know what should be restored, for example:

- You have a post with 3 comments
- You delete a comment
- Then delete the post
- Then you decide you need to restore the post, should we now restore all the comments?

If we restore all of them then obviously one is restored which shouldn't, even if we restore the ones with the same deleted_at time, we could run into problems with concurrent sessions. (Session A deleted the comment and session B deleted the post).

And what should we do if there are multiple connections:
- You have a Post, a Comment and an Author
- You delete the Post which deletes the comment
- You then delete the author which should also remove the already deleted comment
- Then you restore the post which tries to restore the comment, now you somehow need to know that the comment is restorable or not.

I'm not trying to say that this is a bad idea, just that it is a complex one.

**Reply by om3rcitak** [reply]
Yes, absolutely this feature is very complex. And your questions is very make sense. I think so if we add `'deleted_by'` column too in `$table->softDeletes();` method, we are fix many problem. I didn't sure, I just writing during thinking.
